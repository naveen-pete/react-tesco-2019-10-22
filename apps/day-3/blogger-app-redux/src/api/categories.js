const apiUrl = 'http://localhost:3001/categories';

export const getCategories = () => {
  return fetch(apiUrl)
    .then(response => response.json());
};

export const getCategory = id => {
  return fetch(`${apiUrl}/${id}`)
    .then(response => response.json());
};
