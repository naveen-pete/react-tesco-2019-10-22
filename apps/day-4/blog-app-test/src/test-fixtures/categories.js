const categories = [
  {
    "name": "React",
    "code": "react"
  },
  {
    "name": "Redux",
    "code": "redux"
  },
  {
    "name": "Angular",
    "code": "angular"
  },
  {
    "name": "ES6",
    "code": "es6"
  }
];

export default categories;
