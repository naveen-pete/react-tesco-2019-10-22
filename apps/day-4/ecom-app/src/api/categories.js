const apiUrl = 'http://localhost:3000/api/categories';

export const getCategories = async () => {
  const response = await fetch(apiUrl);
  return response.json();
};
