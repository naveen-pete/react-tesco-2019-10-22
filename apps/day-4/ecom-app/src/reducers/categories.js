
const categoriesReducer = (state = [], action) => {
  switch (action.type) {
    case 'GET_CATEGORIES':
      return [...action.payload];

    default:
      return state;
  };
};

export default categoriesReducer;
