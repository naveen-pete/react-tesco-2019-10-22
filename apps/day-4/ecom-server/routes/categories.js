const express = require('express');
const router = express.Router();

const { Category } = require('../models/category');

router.get('/', async (req, res) => {
  try {
    const categories = await Category.find().select({ __v: 0 });
    res.json(categories);
  } catch (e) {
    res.status(500).json({ message: 'Server error.' });
    console.log('Error:', e.message);
  }
});

router.get('/:id', async (req, res) => {
  try {
    const category = await Category.findById(req.params.id).select({ __v: 0 });

    if (!category) {
      return res.status(404).json({ message: 'Category does not exist!' });
    }

    res.json(category);
  } catch (e) {
    res.status(500).json({ message: 'Server error.' });
    console.log('Error:', e.message);
  }
});

module.exports = router;