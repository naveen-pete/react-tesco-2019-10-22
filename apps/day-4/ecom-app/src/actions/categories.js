import { getCategories as getCategoriesApi } from '../api/categories';

export const getCategories = () => {
  return async dispatch => {
    try {
      const categories = await getCategoriesApi();
      dispatch({
        type: 'GET_CATEGORIES',
        payload: categories
      });
    } catch (e) {
      console.log('Get categories failed!')
      console.log('Error:', e);
    }
  };
}