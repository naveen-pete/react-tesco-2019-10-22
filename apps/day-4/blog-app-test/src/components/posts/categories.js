import React, { Component } from 'react';
import FontAwesome from 'react-fontawesome';

import { connect } from 'react-redux';

import { CATEGORY_ALL } from './defaults';
import { getCategories, setCurrentCategory } from '../../actions/posts';

export class Categories extends Component {
  componentDidMount() {
    this.props.getCategories();
  }

  onCategoryClick(selectedCategory) {
    this.props.setCurrentCategory(selectedCategory);
  }

  addActiveIfSelected(category) {
    const { code } = this.props.currentCategory;
    return code === category.code ? 'active' : '';
  }

  render() {
    let { categories } = this.props;
    categories = [CATEGORY_ALL, ...categories];

    return (
      <div className="col-sm-3">
        <h4>
          <FontAwesome name="object-group" /> Category
        </h4>
        <ul className="list-group">
          {categories.map(category => (
            <li
              onClick={() => this.onCategoryClick(category)}
              key={category.code}
              className={`list-group-item cursorPointer ${this.addActiveIfSelected(category)}`}
            >
              {category.name}
            </li>
          ))}
        </ul>
      </div>
    );
  }
}

const mapStateToProps = ({ categories, currentCategory }) => ({
  categories,
  currentCategory
});

const mapDispatchToProps = dispatch => ({
  getCategories: () => dispatch(getCategories()),
  setCurrentCategory: category => dispatch(setCurrentCategory(category))
});

export default connect(mapStateToProps, mapDispatchToProps)(Categories);
