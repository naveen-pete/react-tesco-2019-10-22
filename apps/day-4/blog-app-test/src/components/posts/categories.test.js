import React from 'react';
import { shallow } from 'enzyme';
import { Categories } from './categories';
import categories from '../../test-fixtures/categories';

let currentCategory, getCategoriesSpy, setCurrentCategorySpy;

beforeEach(() => {
  currentCategory = {
    code: 'react',
    name: 'React'
  };

  getCategoriesSpy = jest.fn();
  setCurrentCategorySpy = jest.fn();
});

test('should render categories component properly', () => {
  const wrapper = shallow(
    <Categories 
      currentCategory={currentCategory}
      categories={categories} 
      getCategories={getCategoriesSpy}
    />
  );
  expect(wrapper).toMatchSnapshot();
});

test('should render the list of categories', () => {
  const wrapper = shallow(
    <Categories 
      currentCategory={currentCategory}
      categories={categories} 
      getCategories={getCategoriesSpy}
    />
  );

  expect(wrapper).toMatchSnapshot();
  const items = wrapper.find('li');
  expect(items).toHaveLength(categories.length + 1);
});

test('should render the list of categories', () => {
  const wrapper = shallow(
    <Categories 
      currentCategory={currentCategory}
      categories={categories} 
      getCategories={getCategoriesSpy}
    />
  );

  expect(wrapper).toMatchSnapshot();
  const items = wrapper.find('li');
  categories.forEach((category, i) => {
    expect(items.at(i + 1).text()).toBe(category.name);
  });
});

test('should set current category when clicked', () => {
  const wrapper = shallow(
    <Categories 
      currentCategory={currentCategory}
      categories={categories} 
      getCategories={getCategoriesSpy}
      setCurrentCategory={setCurrentCategorySpy}
    />
  );

  expect(wrapper).toMatchSnapshot();
  wrapper.find('li').at(2).simulate('click');
  expect(setCurrentCategorySpy).toHaveBeenCalledWith(categories[1]);
});

test('should set \'active\' css class to currently selected category', () => {
  const wrapper = shallow(
    <Categories 
      currentCategory={currentCategory}
      categories={categories} 
      getCategories={getCategoriesSpy}
      setCurrentCategory={setCurrentCategorySpy}
    />
  );

  expect(wrapper).toMatchSnapshot();
  expect(wrapper.find('li.active').text()).toBe(currentCategory.name);
});
