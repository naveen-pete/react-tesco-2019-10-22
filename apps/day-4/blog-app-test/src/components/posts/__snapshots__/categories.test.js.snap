// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render categories component properly 1`] = `
<div
  className="col-sm-3"
>
  <h4>
    <FontAwesome
      name="object-group"
    />
     Category
  </h4>
  <ul
    className="list-group"
  >
    <li
      className="list-group-item cursorPointer "
      key="all"
      onClick={[Function]}
    >
      All
    </li>
    <li
      className="list-group-item cursorPointer active"
      key="react"
      onClick={[Function]}
    >
      React
    </li>
    <li
      className="list-group-item cursorPointer "
      key="redux"
      onClick={[Function]}
    >
      Redux
    </li>
    <li
      className="list-group-item cursorPointer "
      key="angular"
      onClick={[Function]}
    >
      Angular
    </li>
    <li
      className="list-group-item cursorPointer "
      key="es6"
      onClick={[Function]}
    >
      ES6
    </li>
  </ul>
</div>
`;

exports[`should render the list of categories 1`] = `
<div
  className="col-sm-3"
>
  <h4>
    <FontAwesome
      name="object-group"
    />
     Category
  </h4>
  <ul
    className="list-group"
  >
    <li
      className="list-group-item cursorPointer "
      key="all"
      onClick={[Function]}
    >
      All
    </li>
    <li
      className="list-group-item cursorPointer active"
      key="react"
      onClick={[Function]}
    >
      React
    </li>
    <li
      className="list-group-item cursorPointer "
      key="redux"
      onClick={[Function]}
    >
      Redux
    </li>
    <li
      className="list-group-item cursorPointer "
      key="angular"
      onClick={[Function]}
    >
      Angular
    </li>
    <li
      className="list-group-item cursorPointer "
      key="es6"
      onClick={[Function]}
    >
      ES6
    </li>
  </ul>
</div>
`;

exports[`should render the list of categories 2`] = `
<div
  className="col-sm-3"
>
  <h4>
    <FontAwesome
      name="object-group"
    />
     Category
  </h4>
  <ul
    className="list-group"
  >
    <li
      className="list-group-item cursorPointer "
      key="all"
      onClick={[Function]}
    >
      All
    </li>
    <li
      className="list-group-item cursorPointer active"
      key="react"
      onClick={[Function]}
    >
      React
    </li>
    <li
      className="list-group-item cursorPointer "
      key="redux"
      onClick={[Function]}
    >
      Redux
    </li>
    <li
      className="list-group-item cursorPointer "
      key="angular"
      onClick={[Function]}
    >
      Angular
    </li>
    <li
      className="list-group-item cursorPointer "
      key="es6"
      onClick={[Function]}
    >
      ES6
    </li>
  </ul>
</div>
`;

exports[`should set 'active' css class to currently selected category 1`] = `
<div
  className="col-sm-3"
>
  <h4>
    <FontAwesome
      name="object-group"
    />
     Category
  </h4>
  <ul
    className="list-group"
  >
    <li
      className="list-group-item cursorPointer "
      key="all"
      onClick={[Function]}
    >
      All
    </li>
    <li
      className="list-group-item cursorPointer active"
      key="react"
      onClick={[Function]}
    >
      React
    </li>
    <li
      className="list-group-item cursorPointer "
      key="redux"
      onClick={[Function]}
    >
      Redux
    </li>
    <li
      className="list-group-item cursorPointer "
      key="angular"
      onClick={[Function]}
    >
      Angular
    </li>
    <li
      className="list-group-item cursorPointer "
      key="es6"
      onClick={[Function]}
    >
      ES6
    </li>
  </ul>
</div>
`;

exports[`should set current category when clicked 1`] = `
<div
  className="col-sm-3"
>
  <h4>
    <FontAwesome
      name="object-group"
    />
     Category
  </h4>
  <ul
    className="list-group"
  >
    <li
      className="list-group-item cursorPointer "
      key="all"
      onClick={[Function]}
    >
      All
    </li>
    <li
      className="list-group-item cursorPointer active"
      key="react"
      onClick={[Function]}
    >
      React
    </li>
    <li
      className="list-group-item cursorPointer "
      key="redux"
      onClick={[Function]}
    >
      Redux
    </li>
    <li
      className="list-group-item cursorPointer "
      key="angular"
      onClick={[Function]}
    >
      Angular
    </li>
    <li
      className="list-group-item cursorPointer "
      key="es6"
      onClick={[Function]}
    >
      ES6
    </li>
  </ul>
</div>
`;
